### PR TITLE
feat: Otto contract

### PR DESCRIPTION
**Pull request by d0c0q**

### CODE REVIEW

File: contracts/Otto.sol [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    function setName(uint256 tokenId_, string memory name_)
+        external
+        validOttoId(tokenId_)
+        onlyOttoOwner(tokenId_)
+    {
+        infos[tokenId_].name = name_;
+    }
+
+    function setDescription(uint256 tokenId_, string memory desc_)
```
[the remark this answers is not in the document]
Comment (d0c0q) [reply contracts/Otto.sol]: https://github.com/OtterClam/otter-contracts/pull/13/commits/c8f7b7e0e57524653dd647a505f4fea333789f37

File: contracts/Otto.sol [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    function grantOperator(address operator_) public onlyAdmin {
+        _setupRole(OPERATOR_ROLE, operator_);
+    }
+
+    function revokeOperator(address operator_) public onlyAdmin {
+        _revokeRole(OPERATOR_ROLE, operator_);
+    }
+
+    function setName(uint256 tokenId_, string memory name_)
```
[the remark this answers is not in the document]
Comment (d0c0q) [reply contracts/Otto.sol]: https://github.com/OtterClam/otter-contracts/pull/13/commits/c8f7b7e0e57524653dd647a505f4fea333789f37

File: contracts/Otto.sol [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        string memory name_,
+        string memory symbol_,
+        uint256 maxBatchSize_,
+        uint256 collectionSize_
+    ) public virtual override initializer {
+        ERC721AUpgradeable.initialize(
+            name_,
+            symbol_,
+            maxBatchSize_,
+            collectionSize_
```
[the remark this answers is not in the document]
Comment (d0c0q) [reply contracts/Otto.sol:74]: https://github.com/OtterClam/otter-contracts/pull/13/commits/3af418392ce9f9ccc14c75319816f3266783f65d

File: contracts/Otto.sol [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        string description;
+        uint256 birthday;
+        uint256 traits;
+        uint256 level;
+        uint256 experiences;
+        uint256 hungerValue;
+        uint256 friendship;
+        // [STR, DEF, DEX, INT, LUK, CON, CUTE, BRS]
+        int16[NUM_ATTRIBUTES] attributes; // can be changed by level up
+        int16[NUM_ATTRIBUTES] attributeBonuses; // from traits & wearable
```
Comment (otterclamking) [inline contracts/Otto.sol:27]: We should make attributes a dynamic array and can be added more in the future.

File: contracts/Otto.sol [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        view
+        virtual
+        override(ERC721AUpgradeable, AccessControlUpgradeable)
+        returns (bool)
+    {
+        return super.supportsInterface(interfaceId);
+    }
+}
+
+contract OttoV2 is Otto {
```
Comment (otterclamking) [inline contracts/Otto.sol:218]: No need to keep this

File: contracts/OttopiaPortalCreator.sol [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        uint256 clamBalance = CLAM.balanceOf(address(this));
+        CLAM.transfer(dao, clamBalance);
+
+        uint256 wethBalance = WETH.balanceOf(address(this));
+        uint256 toTreasury = _calcPercentage(wethBalance, 5000);
+        WETH.transfer(treasury, toTreasury);
+        WETH.transfer(dao, wethBalance - toTreasury);
+    }
+
+    function giveaway(address to_, uint256 quantity_) external onlyOwner {
```
Comment (otterclamking) [inline contracts/OttopiaPortalCreator.sol:115]: Use the name `devMint`

File: contracts/OttopiaPortalCreator.sol [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    function stopSale() external onlyOwner {
+        saleStage = SALE_STAGE.NOT_STARTED;
+    }
+
+    function startPreSale() external onlyOwner {
+        saleStage = SALE_STAGE.PRE_SALE;
+    }
+
+    function startPublicSale() external onlyOwner {
+        saleStage = SALE_STAGE.PUBLIC_SALE;
```
Comment (otterclamking) [inline contracts/OttopiaPortalCreator.sol:97]: need to use timestamp to decide the phase

File: contracts/Otto.sol [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+import './libraries/ERC721AUpgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol';
+
+contract Otto is ERC721AUpgradeable, AccessControlUpgradeable, IOtto {
+    bytes32 public constant MINTER_ROLE = keccak256('MINTER_ROLE');
+    bytes32 public constant MANAGER_ROLE = keccak256('MANAGER_ROLE');
+    uint256 public constant NUM_ATTRIBUTES = 8;
+
+    string private _baseTokenURI;
+    mapping(uint256 => OttoInfo) private infos;
```
Comment (otterclamking) [inline contracts/Otto.sol:14]: public

File: contracts/Otto.sol (hunk 4 above)
Comment (d0c0q) [reply contracts/Otto.sol]: https://github.com/OtterClam/otter-contracts/pull/13/commits/37a7c87d30a61e05b55596e2ebd7209c4b04f57d

File: contracts/Otto.sol (hunk 8 above)
Comment (d0c0q) [reply contracts/Otto.sol]: https://github.com/OtterClam/otter-contracts/pull/13/commits/9664e08a7ef2977748351bf5accb62d996fe483f

File: contracts/OttopiaPortalCreator.sol (hunk 6 above)
Comment (d0c0q) [reply contracts/OttopiaPortalCreator.sol]: https://github.com/OtterClam/otter-contracts/pull/13/commits/6bfca76d060cf4e41ceef0b2a47c343ef3c54e53

File: contracts/OttopiaPortalCreator.sol (hunk 7 above)
Comment (d0c0q) [reply contracts/OttopiaPortalCreator.sol]: https://github.com/OtterClam/otter-contracts/pull/13/commits/72d4304b4185a34fe6f8a1a0a0aca652673b9e85

File: contracts/Otto.sol (hunk 5 above)
Comment (d0c0q) [reply contracts/Otto.sol]: https://github.com/OtterClam/otter-contracts/pull/13/commits/82ff8d095bcb30683cc4eedc4f14b7c8bc83d77e

File: contracts/Otto.sol [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        pure
+        returns (uint8[32] memory arr_)
+    {
+        for (uint8 i = 0; i < 32; i++) {
+            arr_[i] = uint8(traits_ >> (i * 8));
+        }
+    }
+}
+
+// contract OttoV2 is Otto {
```
Comment (otterclamking) [inline contracts/Otto.sol:238]: remove all comments, don't keep in code

Comment (d0c0q) [reply contracts/Otto.sol]: https://github.com/OtterClam/otter-contracts/pull/13/commits/da5b108f400251ca2ffcbbcabefb48752fe956e9

**Review by otterclamking — APPROVED** [review]

**Pull request merged**